File: ss_binary_search.py

def binary_search(arr, x):
	start = 0
	end = len(arr) - 1
	while start <= end:
		middle = (start + end) / 2
		if arr[middle] > x:
			end = middle - 1
		elif arr[middle] < x:
			start = middle + 1
		else:
			return middle
	return -1

def binary_search_twisted(arr, x):
	orig_start = 0
	prev = arr[0]
	i = 1
	l_arr = len(arr)
	while i < l_arr:
		if arr[i] < prev:
			break
		prev = arr[i]
		i += 1
	orig_start = i

	start = 0
	end = len(arr) - 1
	while start <= end:
		middle = (start + end) / 2
		if arr[(middle + orig_start) % l_arr] > x:
			end = middle - 1
		elif arr[(middle + orig_start) % l_arr] < x:
			start = middle + 1
		else:
			return (middle + orig_start) % l_arr
	return -1

def main():
	arr = [i for i in xrange(200, 250)] + [i for i in xrange(200)]
	print(arr)
	print(binary_search_twisted(arr, 210))

if __name__ == "__main__":
	main()

File: ss_binary_search.cpp
#include <iostream>

#define NULL_VALUE -1
#define SIZE 100

using namespace std;

int binarySearch(int* arr, int x){
	int start = 0;
	int end = SIZE - 1;
	while(start <= end){
		int middle = (start + end) / 2;
		if(arr[middle] > x){
			end = middle - 1;
		}
		else if(arr[middle] < x){
			start = middle + 1;
		}
		else{
			return middle;
		}
	}
	return NULL_VALUE;
}

int binarySearchTwisted(int* arr, int x){
	int origStart = 0;
	int middle = 0;
	int prev = arr[0];
	int i = 1;
	int l_arr = SIZE;
	while(i < l_arr){
		if(arr[i] < prev){
			break;
		}
		prev = arr[i];
		i += 1;
	}
	origStart = i;

	int start = 0;
	int end = SIZE - 1;
	while(start <= end){
		middle = (start + end) / 2;
		if(arr[(middle + origStart) % l_arr] > x){
			end = middle - 1;
		}
		else if(arr[(middle + origStart) % l_arr] < x){
			start = middle + 1;
		}
		else{
			return (middle + origStart) % l_arr;
		}
	}
	return NULL_VALUE;
}

int main(){

	int arr[SIZE];
	for(int i=50; i<SIZE; i++){
		arr[i] = i;
		cout << arr[i] << endl;
	}
	for(int i=0; i<50; i++){
		arr[i] = i;
		cout << arr[i] << endl;
	}

	cout << "binarySearchTwisted" << endl;

	cout << binarySearchTwisted(arr, 100) << endl;

	return 0;
}
